Lib/copy.py

```diff
@@ -126,16 +126,20 @@ def deepcopy(x, memo=None, _nil=[]):
     if cls in _atomic_types:
         return x
 
-    d = id(x)
     if memo is None:
+        if cls in _builtin_iterables and not x:
```

Comment:
What will happen if we move this check before `if memo is None:`? How will the timings change?

It will simplify the code quite a lot.

Reply:
Moving this check to before `if memo is None:` should make it faster for all cases.

We can easily do this for `tuple` and `frozenset` since they are immutable so we don't need to create a copy similar to the existing atomic type fast path.
However, for the mutable case moving the check outside the if statement would actually change the behaviour. Currently when `memo` is passed into the function we'd return the cached value instead of creating a new one. I think we need to keep this behaviour also for empty containers of am I missing something?

Reply:
Hm. Documentation says:

> copy.deepcopy(obj[, memo])
> Return a deep copy of obj.

Plus:

> Two problems often exist with deep copy operations that don’t exist with shallow copy operations:
> - Recursive objects (compound objects that, directly or indirectly, contain a reference to themselves) may cause a recursive loop.
> - Because deep copy copies everything it may copy too much, such as data which is intended to be shared between copies.

> The [deepcopy()](https://docs.python.org/dev/library/copy.html#copy.deepcopy) function avoids these problems by:
> - keeping a memo dictionary of objects already copied during the current copying pass; and
> - letting user-defined classes override the copying operation or the set of components copied.

We don't make any promises about the object's origin. So, you can consider `memo`'s contents as a solution for these two problems above. 

So, I don't see a problem with returning a new copy of an empty std container.

Reply:
@sobolevn Doesn't your suggestion *"copy too much, such as data which is intended to be shared between copies"*?

Reply:
> So, I don't see a problem with returning a new copy of an empty std container.

This becomes a problem with nested containers that are expected to share data. Consider the following example:
```python
import copy

l = []
d = {"foo": l, "bar": l}
d_copy = copy.deepcopy(d)

d["foo"].append(1)
assert d == {'foo': [1], 'bar': [1]}

d_copy["foo"].append(1)
assert d_copy == {'foo': [1], 'bar': [1]}
```
Now we'd like `copy.deepcopy(d)` to behave the same way, which is only the case if we respect `memo` as it is passed internally when copying nested containers. This example would fail in case we return a new empty list everytime.